### PR TITLE
[PT2] weight_compression

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
### Changes

Implement weight compression algorithms for experimental PT tracing

### Related tickets

152996

### CODE REVIEW

**Comment by AlexanderDokuchaev** [comment]
Model | Metric value | Metric diff | Num int8 | Num int4 | Compr. time | Stat. collection time | Mixed-Precision search time | Apply Compression time | Total time | RAM MiB | RAM MiB System | Status
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
tinyllama_data_aware | 0.85767 | -0.14233 | 123 | 94 | 0:01:00 | 0:00:51 | 0:00:04 | 0:00:02 | 0:03:09 | 5099 | 1618 | XFAIL: num_compressed_xfail_reason - Regression: The number of int8 ops   is different than reference 124 != 123
tinyllama_int8_data_free | 0.95624 | -0.04376 | 311 | 0 | 0:00:03 |   |   | 0:00:02 | 0:02:25 | 5336 | 2499 | XFAIL: num_compressed_xfail_reason - Regression: The number of int8 ops   is different than reference 312 != 311
tinyllama_int4_data_free | 0.73873 | -0.26127 | 83 | 114 | 0:00:08 |   | 0:00:04 | 0:00:02 | 0:02:16 | 5201 | 1990 | XFAIL: num_compressed_xfail_reason - Regression: The number of int8 ops   is different than reference 84 != 83
tinyllama_scale_estimation_per_channel | 0.80799 | -0.19201 | 123 | 188 | 0:02:31 | 0:00:55 | 0:00:04 | 0:00:01 | 0:04:40 | 5280 | 4012 | XFAIL: num_compressed_xfail_reason - Regression: The number of int8 ops   is different than reference 124 != 123
tinyllama_scale_estimation_group_size_64 | 0.8566 | -0.1434 | 123 | 94 | 0:03:21 | 0:00:57 | 0:00:04 | 0:00:02 | 0:05:29 | 5461 | 4287 | XFAIL: num_compressed_xfail_reason - Regression: The number of int8 ops   is different than reference 124 != 123

File: nncf/quantization/algorithms/weight_compression/torch_backend.py [hunk 1]
```diff
@@ -291,38 +318,55 @@ def transform_model(
...
-
-            # inserts the weight decompressor into the model as the post hook on the model weight
-            transformation_layout.register(
-                PTSharedFnInsertionCommand(
-                    [PTTargetPoint(TargetType.OPERATOR_POST_HOOK, target_node_name=weight_node.node_name)],
-                    decompressor,
-                    decompressor_name,
+            if is_experimental_torch_tracing_enabled():
+                weight.requires_grad = False
+                weight.data = packed_tensor
```
Comment (daniil-lyakhov) [inline nncf/quantization/algorithms/weight_compression/torch_backend.py:331]: What do you think about using a `break` statement here to reduce amount of indents?

**Review by daniil-lyakhov — APPROVED** [review]
Minor

File: nncf/quantization/algorithms/weight_compression/torch_backend.py [hunk 2]
```diff
@@ -291,38 +318,55 @@ def transform_model(
...
+                        [
+                            PTTargetPoint(
+                                TargetType.OPERATOR_POST_HOOK, target_node_name=weight_node.node_name.replace(".", ":")
+                            )
+                        ],
+                        decompressor,
+                    )
+                )
+            else:
+                compressed_parameter = torch.nn.Parameter(packed_tensor, requires_grad=False)
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/quantization/algorithms/weight_compression/torch_backend.py:344]: No found any diff in memory usage

```pyhton
import time
import torch
from torch import nn

class LargeConvModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(1024, 1024, 5)
    def forward(self, x):
        return self.conv(x)

model = LargeConvModel()
input_tensor = torch.randn(1, 1024, 32, 32)

for x in range(100):
    model.conv.weight = nn.Parameter(torch.randn(1024, 1024, 3, 3, dtype=torch.float32), requires_grad=False)
    model(input_tensor)
    time.sleep(0.1)
```
image
image



```python
for x in range(100):
    model.conv.weight.requires_grad = False
    model.conv.weight.data = torch.randn(1024, 1024, 3, 3, dtype=torch.float32)
    model(input_tensor)
    time.sleep(0.1)
```

image
image

**Review by alexsu52 — COMMENTED** [review]
LGTM

I run WC: https://github.com/openvinotoolkit/nncf/actions/runs/13714753948

**Review by alexsu52 — APPROVED** [review]
